tests/conftest.py

```diff
@@ -355,3 +408,148 @@ def fake_fgs_composite(smargon: Smargon, test_fgs_params: InternalParameters):
+        Args:
+             document_name: document name as defined in the Bluesky Message Protocol 'subscribe' call,
+             all subscribers filtering on this document name will be called
+             document: the document to send
+        """
+        for callback_func, callback_docname in self.callbacks.values():
+            if callback_docname == "all" or callback_docname == document_name:
+                callback_func(document_name, document)
+
+    def simulate_plan(self, gen: Generator[Msg, object, object]) -> list[Msg]:
+        """Simulate the RunEngine executing the plan
+        Args:
+            gen: the generator function that executes the plan
+        Returns:
+            a list of the messages generated by the plan
+        """
+        messages = []
+        send_value = None
+        try:
+            while msg := gen.send(send_value):
+                send_value = None
+                messages.append(msg)
+                LOGGER.debug(f"<{msg}")
+                if handler := next(
+                    (h for h in self.message_handlers if h.predicate(msg)), None
+                ):
+                    send_value = handler.runnable(msg)
+
+                if send_value:
+                    LOGGER.debug(f">send {send_value}")
+        except StopIteration:
+            pass
+        return messages
+
+    def _add_callback(self, msg_args):
+        self.callbacks[self.next_callback_token] = msg_args
+        self.next_callback_token += 1
+
+    def assert_message_and_return_remaining(
+        self, messages: list[Msg], predicate: Callable[[Msg], bool], group: str = None
+    ):
+        """Find the next message matching the predicate, assert that we found it
+        Return: all the remaining messages starting from the matched message"""
+        indices = [
+            i
+            for i in range(len(messages))
+            if (
+                not group
+                or (messages[i].kwargs and messages[i].kwargs.get("group") == group)
+            )
+            and predicate(messages[i])
+        ]
+        assert indices, f"Nothing matched predicate {predicate}"
+        return messages[indices[0] :]
+
+    def mock_message_generator(
+        self,
+        function_name: str,
+    ) -> Callable[..., Generator[Msg, object, object]]:
+        """Returns a callable that returns a generator yielding a Msg object recording the call arguments.
+        This can be used to mock methods returning a bluesky plan or portion thereof, call it from within a unit test
+        using the RunEngineSimulator, and then perform asserts on the message to verify in-order execution of the plan"""
+
+        def mock_method(*args, **kwargs):
+            yield Msg(function_name, None, *args, **kwargs)
+
+        return mock_method
+
+
+class MessageHandler:
+    def __init__(self, p: Callable[[Msg], bool], r: Callable[[Msg], object]):
+        self.predicate = p
+        self.runnable = r
+
+
+@pytest.fixture
+def sim():
```

Comment:
Sorry, can I raise this to a should after seeing it used a bit more? I think `sim_run_engine` would be clearer